dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/HibernateTrackedEntityStore.java

```diff
@@ -318,9 +303,7 @@ private String getCountQueryWithMaxTrackedEntityLimit(TrackedEntityQueryParams p
         + getQuerySelect(params)
         + "FROM "
         + getFromSubQuery(params, true, null)
-        + (params.getEnrolledInTrackerProgram().getMaxTeiCountToReturn() > 0
-            ? getLimitClause(params.getEnrolledInTrackerProgram().getMaxTeiCountToReturn() + 1)
-            : "")
+        + (params.getMaxTeLimit() > 0 ? getLimitClause(params.getMaxTeLimit() + 1) : "")
```

Comment:
I think we can optimize this further. This count method would/should only be called if there is maxTeLimit >0 . So either we start this method with something like 

```java
if (!params.hasMaxTeLimit())
return 1;
``` 
We do not call this method if maxTeLimit is 0, but this is just to prevent another caller from calling this method maybe. 

Once that is done, then we can safely guarantee the existence of the limit in the params and just use that. And avoid this ternary expression.

Reply:
@ameenhere 

This count is only used if `totalPages=true` to show the count in the pager. Why do we need the +1 in that case? I also don't get why we use `params.getMaxTeLimit()` here 🤔 

If I get this: `params.getMaxTeLimit()` is set if there is a TET/Program limit. The check that the users search is within this limit is already done in the service.

Should the `systemTeLimit` not also be considered here? We did validate that the `pageSize` <= `systemTeLimit` in the service but the count can exceed that

Reply:
This is the count used by the service to verify we are not over the limit of the program or TET when searching outside the capture scope.

In the service, there's the validation `maxTeLimit>0`, and since I think of the store as a class that only the tracker team uses, I didn't think about adding the same guard clause there, but I'm open to it if you feel it's beneficial.

@teleivo can it be you are referring to `getCountQuery(TrackedEntityQueryParams params)`?

Reply:
> @teleivo can it be you are referring to getCountQuery(TrackedEntityQueryParams params)?

you are right @muilpp , my bad 😅 

You are also right that the store is only used by our service. So we don't need to redo any validation as long as we ensure the inputs to the store are safe to use from all code paths 👌🏻

Reply:
Please look at the flow (maybe debug). The params limit is set after this method is even reached AFAIK. So we would never actually add a `limit` to the SQL query.

Reply:
Yes the maxTeLimit obtained (from Program/TET) is not set into the params early enough.